server/catgenome/src/main/java/com/epam/catgenome/manager/externaldb/homologene/HomologeneManager.java

```diff
@@ -195,28 +191,127 @@ private EntryGenesXML getEntryGenes(final Document doc) {
-            spf.setFeature("http://xml.org/sax/features/external-general-entities", false);
-            spf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
-            spf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
-
-            XMLReader xmlReader = spf.newSAXParser().getXMLReader();
-            InputSource inputSource = new InputSource(path);
-            SAXSource source = new SAXSource(xmlReader, inputSource);
-            Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
-            HomologeneEntrySetXML homologeneEntrySet = (HomologeneEntrySetXML) unmarshaller.unmarshal(source);
-            return homologeneEntrySet.getHomologeneEntrySetEntries().getHomologeneEntries();
-        } catch (JAXBException | ParserConfigurationException | SAXException e) {
-            log.error(e.getMessage());
-            return Collections.emptyList();
+    @SneakyThrows
+    public List<HomologeneEntry> readHomologenes(final String path) {
+        XMLInputFactory xmlInputFactory = XMLInputFactory.newInstance();
+        xmlInputFactory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+        InputStream inputStream = new FileInputStream(path);
+        XMLStreamReader streamReader = xmlInputFactory.createXMLStreamReader(inputStream);
+        List<HomologeneEntry> homologeneEntries = new ArrayList<>();
+        List<Gene> genes = new ArrayList<>();
+        List<String> aliases = new ArrayList<>();
+        List<Domain> domains = new ArrayList<>();
+        HomologeneEntry homologeneEntry = null;
+        Gene gene = null;
+        Domain domain = null;
+        while(streamReader.hasNext()) {
+            streamReader.next();
+            if (streamReader.getEventType() == XMLStreamReader.START_ELEMENT) {
+                switch (streamReader.getLocalName()) {
+                    case "HG-Entry":
+                        homologeneEntry = new HomologeneEntry();
+                        break;
+                    case "HG-Entry_hg-id":
+                        homologeneEntry.setGroupId(Long.valueOf(streamReader.getElementText()));
+                        break;
+                    case "HG-Entry_version":
+                        homologeneEntry.setVersion(Long.valueOf(streamReader.getElementText()));
+                        break;
+                    case "HG-Entry_caption":
+                        homologeneEntry.setCaption(streamReader.getElementText());
+                        break;
+                    case "HG-Entry_taxid":
+                        homologeneEntry.setTaxId(Long.valueOf(streamReader.getElementText()));
+                        break;
+                    case "HG-Gene":
+                        gene = new Gene();
+                        break;
+                    case "HG-Gene_geneid":
+                        gene.setGeneId(Long.valueOf(streamReader.getElementText()));
```

Comment:
I would do here and for each other situation for `gene`  `homologeneEntry`  `Domain`  some check for null for `gene`, and thorw an IlligalStateExc if it is null, I understand it shouldn't be possible but still

you could write some additional method that will firstly check for null an object and after that perform Consumer function that will do actual setting